sparse/client.go

```diff
@@ -178,80 +132,121 @@ func (session *SyncSessionClient) syncFileContent(file FileIoProcessor, fileSize
 		log.Debugf("The file is a hole: %s", holeInterval)
 
 		// syncing hole interval
-		err := session.SyncHoleInterval(holeInterval)
+		err := client.syncHoleInterval(holeInterval)
 		if err != nil {
-			log.Debugf("SyncHoleInterval failed: %s", holeInterval)
+			log.Errorf("syncHoleInterval failed: %s", holeInterval)
 			return err
 		}
 	}
 
-	return session.send(requestHeader{requestMagic, syncDone})
+	return client.closeServer()
 }
 
-func (session *SyncSessionClient) sendSyncRequest(path string, size int64) error {
-	err := session.send(requestHeader{requestMagic, syncRequestCode})
-	if err != nil {
-		log.Debug("send requestHeader failed: ", err)
-		return err
+func (client *syncClient) sendHTTPRequest(method string, action string, interval Interval, data []byte, filePath string) (*http.Response, error) {
+	httpClient := &http.Client{Timeout: time.Duration(httpClientTimeout * time.Second)}
+
+	url := fmt.Sprintf("http://%s:%d/v1-ssync/%s", client.remote.Host, client.remote.Port, action)
+	log.Debugf("method: %s, url: %s, data[] len: %d", method, url, len(data))
+
+	var req *http.Request
+	if data != nil {
+		req, _ = http.NewRequest(method, url, bytes.NewBuffer(data))
+	} else {
+		req, _ = http.NewRequest(method, url, nil)
+	}
+
+	req.Header.Add("Accept", "application/json")
+
+	q := req.URL.Query()
+	q.Add("begin", strconv.FormatInt(interval.Begin, 10))
+	q.Add("end", strconv.FormatInt(interval.End, 10))
+	if filePath != "" {
+		q.Add("filePath", filePath)
 	}
-	err = session.send(path)
+	req.URL.RawQuery = q.Encode()
+
+	log.Debugf("final url with query string: %s", req.URL.String())
+
+	return httpClient.Do(req)
+}
+
+func (client *syncClient) openServer(remotePath string) error {
+	var err error
+	var resp *http.Response
+
+	timeStart := time.Now()
+	timeStop := timeStart.Add(time.Duration(client.timeout) * time.Second)
+	for timeNow := timeStart; timeNow.Before(timeStop); timeNow = time.Now() {
```

Comment:
This statement is weird. How it that different with retry with 1 second in between? I also see the comment below. Do you know first 1 second need to be removed to speed up test? It does sound like a trick.